collections/_tutorials/2022-06-20-tutorial_transformer.md

```diff
@@ -0,0 +1,124 @@
+---
+layout: post
+title:  "The transformer paradigm demystified"
+author: 'Olivier Bernard'
+date:   2022-06-20
+categories: transformer, encoder
+---
+
+# Notes
+
+* Here is a link to an interesting video to understand the transformer theory: [video](https://www.youtube.com/watch?v=YAgjfMR9R_M)
+
+&nbsp;
+
+- [**Introduction**](#introduction)
+- [**Tokenization process**](#tokenization-process)
+- [**Transformer encoder**](#transformer-encoder)
+  - [Norm step](#norm-step)
+  - [MLP step](#mlp-step)  
+  - [Multi-Head Attention block](#multi-head-attention-block)  
+  - [Self-Attention module](#self-attention-module)    
+
+&nbsp;
+
+## **Introduction**
+Transformer paradigm has been first successfully applied in Natural Langage Processing (NLP) and is currently highly investigated for image processing. 
+
+The basic concept of transformer lies in the generation of self attention maps to guide the decision making process of the underlying architecture. To illsutration the underlying mechanism, we will open the black box of the famous [Vision Transformer (ViT)](https://creatis-myriad.github.io/2022/06/01/VisionTransformer.html) method. The overall architecture is given below:
+
+![](/collections/images/transformers/vit_overview.jpg)
+
+&nbsp;
+
+## **Tokenization process**
+The first step is to convert the input data into a sequence of tokens, $$i.e.$$ a sequence of vectors of dimensions $$\mathbb{R}^{1 \times (P^2 C)}$$, with $$P^2 C$$ being the length of each flattened patch. This step is called the tokenization process and involves a simple linear projection ($$i.e.$$ a multiplication with a matrix of dimensions $$\mathbb{R}^{(P^2 C) \times D}$$) and a position embedding step which encodes spatial information. The same linear projection and position embedding are shared to encode each patch. This process is modeled as follows: 
+
+$$z_0 = [x_{class}; \, x^1_p\mathbf{E}; \, x^2_p\mathbf{E}; \, \cdots; \, x^N_p\mathbf{E}] + \mathbf{E}_{pos}$$
+
+with <br>
+&nbsp; &nbsp; &nbsp; &nbsp;$$\mathbf{E}$$ being the linear projection matrix of size $$\mathbf{E} \in \mathbb{R}^{(P^2 \cdot C) \times D}$$ <br>
+&nbsp; &nbsp; &nbsp; &nbsp;$$\mathbf{E}_{pos}$$ being the output of the position embedding operation with dimensions $$\mathbf{E}_{pos} \in \mathbb{R}^{(N+1) \times D}$$
+
+&nbsp;
+
+The figure below illustrates the tokenization process used in ViT.
+
+![](/collections/images/transformers/vit_tokenization.jpg)
+
+&nbsp;
+
+## **Transformer encoder**
+
+The figure below presents an overview of the encoding layer of transformer. It is composed of two main steps.
+
+![](/collections/images/transformers/vit_encoding_layer.jpg)
+
+&nbsp;
+
+The goal of the first step is to compute attention maps between tokens. The output corresponds to $$z^{'}_l$$ which is modeled as follows:
+
+$$z^{'}_l = MSA(LN(z_{l-1})) + z_{l-1} \quad \quad \quad \quad l=1 \cdots L$$
```

Comment:
Do you mean MHA here? For _Multi-Head Attention_? Because I don't know what MSA would stand for.

Also a few typesetting suggestions to improve the readability.
```suggestion
$$z^{'}_l = \text{MHA}(\text{LN}(z_{l-1})) + z_{l-1} \quad \quad \quad \quad l=1 \cdots L$$
```